application/src/main/java/org/opentripplanner/framework/csv/OtpCsvReader.java

```diff
@@ -0,0 +1,114 @@
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.opentripplanner.datastore.api.DataSource;
+import org.opentripplanner.framework.csv.parser.AbstractCsvParser;
+import org.opentripplanner.utils.logging.ProgressTracker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Use this class to read in a CSV file from a {@link DataSource}. You must provide (required):
+ * <ul>
+ *   <li>the {@code datasource} - the csv file to read</li>
+ *   <li>the {@code logger} - the logger ro write progress tracking events.</li>
+ *   <li>the {@code parserFactory} the factory to create a CSV parser for mapping into the row type.</li>
+ *   <li>the {@code rowHandler} a handler for each row read. </li>
+ * </ul>
+ *
+ * The class is responsible for processing the CSV file and orchestrating the parsing process.
+ * A progress tracker is created and will be used to track the reading of the datasource.
+ *
+ * @param <T> The row type.
+ */
+public class OtpCsvReader<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OtpCsvReader.class);
+
+  private DataSource dataSource;
+  private Function<CsvReader, AbstractCsvParser<T>> parserFactory;
+  private Consumer<T> rowHandler;
+
+  @Nullable
+  private Consumer<String> progressLogger;
+
+  private OtpCsvReader() {}
+
+  public static <S> OtpCsvReader<S> of() {
+    return new OtpCsvReader<S>();
+  }
+
+  public OtpCsvReader<T> withDataSource(DataSource dataSource) {
+    this.dataSource = dataSource;
+    return this;
+  }
+
+  /**
+   * The logger is optional. If pressent, a {@link ProgressTracker} is created and the
+   * log events is written to the logger.
+   */
+  public OtpCsvReader<T> withProgressLogger(Consumer<String> logger) {
+    this.progressLogger = logger;
+    return this;
+  }
+
+  public OtpCsvReader<T> withParserFactory(
+    Function<com.csvreader.CsvReader, AbstractCsvParser<T>> parserFactory
+  ) {
+    this.parserFactory = parserFactory;
+    return this;
+  }
+
+  public OtpCsvReader<T> withRowHandler(Consumer<T> rowHandler) {
+    this.rowHandler = rowHandler;
+    return this;
+  }
+
+  public void read() {
+    Objects.requireNonNull(dataSource);
+    Objects.requireNonNull(parserFactory);
+    Objects.requireNonNull(rowHandler);
+    read(rowHandler);
+  }
+
+  private void read(Consumer<T> rowHandler) {
+    ProgressTracker progress = null;
+
+    if (!dataSource.exists()) {
+      return;
+    }
+    if (progressLogger != null) {
+      progress = ProgressTracker.track("Read " + dataSource.name(), 10_000, -1);
+      progressLogger.accept(progress.startMessage());
+    }
+    var reader = new CsvReader(dataSource.asInputStream(), StandardCharsets.UTF_8);
+    var parser = parserFactory.apply(reader);
+
+    if (!parser.headersMatch()) {
+      LOG.error(
+        "The header does not match the expected format for csv file: {}",
+        dataSource.path()
+      );
+      return;
+    }
```

Comment:
This is not working! The EmissionDataReader first try the Route parser, then the trip parser. So, if we have trip emissions this will log a false error.

I will change this to throw a checked exception instead - so the caller need to consider what to do when the headers does not match.